### PR TITLE
Reindex share retry between hit sources

### PR DESCRIPTION
**Pull request by henningandersen**
The client and remote hit sources had each their own retry mechanism,
which would do the same. Supporting resiliency we would have to expand
on the retry mechanisms and as a preparation for that, the retry
mechanism is now shared such that each sub class is only responsible for
sending requests and converting responses/failures to common format.

Part of #42612

### CODE REVIEW

**Comment by henningandersen** [comment]
@elasticmachine update-branch

**Comment by henningandersen** [comment]
@elasticmachine update branch

**Comment by henningandersen** [comment]
Failure looks idential to #43889 but for another test,
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/index/reindex/RejectAwareActionListener.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.function.Consumer;
+
+// public for testing
+public interface RejectAwareActionListener<T> extends ActionListener<T> {
+    void onRejection(Exception e);
+
+    /**
+     * Return a new listener that delegates failure/reject to original but forwards response to responseHandler
+     */
+    default <X> RejectAwareActionListener<X> withResponseHandler(Consumer<X> responseHandler) {
```
Comment (tbrooks8) [inline server/src/main/java/org/elasticsearch/index/reindex/RejectAwareActionListener.java]: My preference is for `default` to be used in a scenario where you are providing default behavior for an interface that some implementation *might* want to overwrite. This feels more like a helper static method such as:

```java
static <X> RejectAwareActionListener<X> withResponseHandler(RejectAwareActionListener<X> errorDelegate, Consumer<X> responseHandler) {
// Implementation
}
```

**Review by Tim-Brooks — APPROVED** [review]
LGTM - I added a comment related to my preferences. But if you feel strongly otherwise, I'm fine with it.

File: server/src/main/java/org/elasticsearch/index/reindex/RejectAwareActionListener.java (hunk 1 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/reindex/RejectAwareActionListener.java]: I agree. Fixed in 07d479303e7b6add9cdb5deea506dafeaa3cd7ed

**Pull request merged**